File: hole_exp/hackable_games/views/auction.py
#!/usr/bin/env python
"""AUCTION -- five lots, ascending price, and money you hold yourself.

The three numbers that decide a bid -- the estimate, what the lot is worth to
you, and the standing high bid -- are one sentence apart in the prompt and
mean nothing next to each other in prose. Drawn as one ladder they are a
picture of the decision: your value at the top, the estimate as the house's
mark, the standing bid climbing towards them.

The bid dial is not capped at the budget. Every bid carries a budget figure
alongside it and the two are separate submissions, so a dial that refused to
go past the budget would be the board making one of them for the player.
"""
from __future__ import annotations

import re
from typing import Optional

from .common import collect

_HEAD = re.compile(r"You are bidder\s+(\d+)\.\s+LOT\s+(\d+)\s*/\s*(\d+),"
                   r"\s+round\s+(\d+)\s*/\s*(\d+)")
_EST = re.compile(r"Estimate\s+(\d+)\.\s+This lot is worth\s+(\d+)\s+to you")
_HIGH0 = re.compile(r"The standing high bid\s+(\d+)\s+\(no bids yet\)\.\s+"
                    r"Minimum raise\s+(\d+)")
_HIGH = re.compile(r"The standing high bid\s+(\d+)\s+from bidder\s+(\d+)\.\s+"
                   r"Minimum raise\s+(\d+)")
_BUDGET = re.compile(r"You start with a budget of\s+(\d+)")
_RESULT = re.compile(r"^lot\s+(\d+):\s+bidder\s+(\d+)\s+won at\s+(\d+)")


def view(phase: str, prompt: str) -> Optional[dict]:
    if phase != "bid":
        return None
    h, e = _HEAD.search(prompt), _EST.search(prompt)
    if not h or not e:
        return None
    zero, high = _HIGH0.search(prompt), _HIGH.search(prompt)
    if zero:
        standing, holder, raise_ = int(zero.group(1)), None, int(zero.group(2))
    elif high:
        standing, holder, raise_ = (int(high.group(1)), int(high.group(2)),
                                    int(high.group(3)))
    else:
        return None
    start = _BUDGET.search(prompt)
    return {
        "kind": "auction_bid",
        "seat": int(h.group(1)), "lot": int(h.group(2)), "lots": int(h.group(3)),
        "round": int(h.group(4)), "rounds": int(h.group(5)),
        "estimate": int(e.group(1)), "worth": int(e.group(2)),
        "standing": standing, "holder": holder, "min_raise": raise_,
        "start_budget": int(start.group(1)) if start else None,
        "results": collect(prompt, _RESULT, lambda m: {
            "lot": int(m.group(1)), "bidder": int(m.group(2)),
            "price": int(m.group(3))}),
        "pass_token": "[bid: pass]",
        "tokens": {"bid": "[bid: {n}]", "budget": "[budget: {n}]"},
    }

File: hole_exp/hackable_games/ui/auction.js
'use strict';
/*
  AUCTION -- five lots, ascending price, money you hold yourself.

  One ladder carries the three numbers the bid turns on: what the lot is worth
  to you at the top, the house's estimate as a mark, and the standing bid
  climbing from the bottom. Read as positions on one scale they are a picture;
  read as three sentences they are three sentences.

  THE REPLY IS ASSEMBLED HERE, NOT TYPED. The adapter hands the board the exact
  shapes the referee reads (`tokens`, `pass_token`); the board fills them and
  sends. Nothing about that wire format reaches the page -- a player who had to
  copy a punctuation pattern out of the prompt would be being measured on
  transcription. Both paths run through the same assembler, so a pass carries
  the same two parts a bid does and cannot come out short by one.

  TWO STATEMENTS, TWO CONTROLS, NEITHER NARROWED. The bid is a dial ranged to
  the lot's value, which the prompt gives you, and its box is not clamped to
  that range -- the prompt states a minimum raise, not a maximum, and a control
  that refused to exceed some number would be inventing a rule the referee
  never stated. The budget is a bare box: unseeded, no floor, no ceiling. The
  only budget figure in the prompt is the one you started with, and ranging the
  control on it would be the board saying the declaration has to agree with it.
  Bid and pass are one row of identical buttons with nothing pre-selected.
*/
window.UI = window.UI || {};

window.UI.auction_bid = function (v, ctx) {
  const K = window.KIT;
  const start = K.memo('auction.budget', v.start_budget);
  // The host substitutes into the adapter's shapes when it offers to; the
  // fallback keeps the board renderable on a plain `{ send }` context.
  const fill = ctx.fill || ((shape, n) => String(shape).replace('{n}', n));
  const box = K.board();

  K.put(box, K.head({
    step: 'Lot ' + v.lot + ' / ' + v.lots,
    title: 'round ' + v.round + ' of ' + v.rounds,
    sub: 'bidder ' + v.seat,
  }));

  const top = Math.max(v.worth, v.estimate, v.standing) * 1.15 || 1;
  const ladder = K.panel('this lot',
    K.meter({
      value: v.standing, max: top, label: 'standing high bid',
      tone: v.standing >= v.worth ? 'bad' : 'good',
      marks: [{ at: v.estimate, label: 'estimate ' + v.estimate },
              { at: v.worth, label: 'worth ' + v.worth + ' to you' }],
      text: v.standing + (v.holder === null ? '  (no bids yet)' : '  from bidder ' + v.holder),
    }),
    K.note('Minimum raise ' + v.min_raise + ', so a raise starts at ' +
      (v.standing + v.min_raise) + '.'));
  if (start !== null) K.put(ladder, K.note('You started with a budget of ' + start + '.'));

  const rows = (v.results || []).map(r => ({
    cls: r.bidder === v.seat ? 'mine' : '',
    cells: ['lot ' + r.lot, 'bidder ' + r.bidder, r.price],
  }));
  const done = K.panel('lots settled',
    rows.length ? K.table(['', 'won by', 'price'], rows) : K.note('none yet'));

  K.put(box, K.panels(ladder, done));

  const bid = K.dial({ lo: 0, hi: Math.max(1, v.worth), value: null, label: 'your bid' });
  // `dial` seeds an absent value to its floor, which would have the board
  // naming an amount the player never named. Clear it back to empty.
  bid.set('');
  const bud = moneyBox('the budget you have left');

  const a = K.act();
  K.put(a, K.el('div', 'bd-panel-h', 'your move on this lot'));
  K.put(a, K.note('Both parts go together: what you are doing about the lot, ' +
    'and the budget you have left.'));

  const controls = K.el('div', 'bd-panels');
  K.put(controls, bid.node, bud.node);
  K.put(a, controls);

  const err = K.note('', 'warn');
  K.put(a, err);

  // One row, one class, prompt order, nothing pre-selected. Each option only
  // has to produce its own opening part; `send` adds the budget both owe.
  const options = [
    {
      label: 'bid',
      lead: () => {
        if (bid.get() === null) { err.textContent = 'name an amount, or pass.'; return null; }
        return fill(v.tokens.bid, bid.get());
      },
    },
    { label: 'pass', lead: () => v.pass_token },
  ];

  const row = K.el('div', 'bd-buttons');
  options.forEach(o => {
    const b = K.el('button', 'bd-opt', o.label);
    b.onclick = () => {
      err.textContent = '';
      const lead = o.lead();
      if (lead === null || lead === undefined) return;
      if (bud.get() === null) {
        err.textContent = 'every reply states the budget you have left, a pass included.';
        return;
      }
      ctx.send(lead + ' ' + fill(v.tokens.budget, bud.get()), 'ui');
    };
    K.put(row, b);
  });
  K.put(a, row);
  K.put(box, a);
  return box;

  function moneyBox(label) {
    // A number box and nothing else. No slider, because a slider needs two
    // ends and the prompt supplies neither for this figure.
    const w = K.el('div', 'bd-dial');
    K.put(w, K.el('div', 'lab', label));
    const inp = K.el('input');
    inp.type = 'number';
    inp.step = 1;
    inp.value = '';
    inp.autocomplete = 'off';
    K.put(w, K.put(K.el('div', 'row'), inp));
    return {
      node: w,
      get: () => (String(inp.value).trim() === '' ? null : Number(inp.value)),
    };
  }
};
